### DISCUSSION TITLE
"Incorrect chunk" error with flatbuffers

### DISCUSSION THREAD

**Discussion by andrewwongagility** [foxglove/mcap#1120] [General] [answered]
I get the following error in Foxglove Studio when recording only PointCloud and transform messages (or only PointCloud messages) to an mcap file:
```
Error: Incorrect chunk CRC 3276123314 (expected 1195241418) [library=libmcap 1.1.0]
    at f.errorWithLibrary (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/2364.js:2:11773)
    at f.loadChunkData (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/2364.js:2:17962)
    at async f.readMessages (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/2364.js:2:15227)
    at async k.messageIterator (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/378.js:1:2030)
    at async y.next (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/5219.js:1:2096)
    at async y.nextBatch (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/5219.js:1:2151)
```
![Screenshot from 2023-07-21 17-11-09](https://github.com/foxglove/community/assets/119512501/f4f0152f-e9c5-4ed0-ba27-a020e43b176c)

I noticed that when I record PointCloud, RawImage, and transform message I did **not** get the same error

![Screenshot from 2023-07-21 17-10-28](https://github.com/foxglove/community/assets/119512501/a40b4222-0544-45c2-8899-005feb09dd4d)

Code for writing pointcloud:
```cpp
struct SerializedFlatBuffer {
  SerializedFlatBuffer() = default;

  SerializedFlatBuffer(const flatbuffers::FlatBufferBuilder& builder);

  const std::byte* data; // pointer to flatbuffer byte data
  size_t size;           // number of bytes in data
};

SerializedFlatBuffer getFoxgloveCloud(std::shared_ptr<PointCloudFrame> cloud_ptr)
{
  // TODO: can this be declared once in .h file, or is it specific to each flatbuffer (seems to be
  // latter)?
  flatbuffers::FlatBufferBuilder builder;
  builder.ForceDefaults(true);

  // number of bytes per element of data vector
  size_t bytes_per_element = sizeof(cloud_ptr->data.at(0));
  size_t data_size = bytes_per_element * cloud_ptr->data.size();

  std::string stream_name = cloud_ptr->stream_name.toString();

  // SensorData::PointCloudFrame does not contain pose, so set all pose fields to 0
  auto dummy_pose = foxglove::CreatePose(builder, foxglove::CreateVector3(builder, 0, 0, 0),
                                         foxglove::CreateQuaternion(builder, 0, 0, 0, 1));

  size_t stride = 3 * bytes_per_element; // At least 3 FLOAT32 fields in each type of pointcloud

  // x is 0 bytes from start of data buffer
  // y is 4 bytes from start of data buffer
  // z is 8 bytes from start of data buffer
  auto fields =
    builder.CreateVector({foxglove::CreatePackedElementField(builder, builder.CreateString("x"), 0,
                                                             foxglove::NumericType::FLOAT32),
                          foxglove::CreatePackedElementField(builder, builder.CreateString("y"), 4,
                                                             foxglove::NumericType::FLOAT32),
                          foxglove::CreatePackedElementField(builder, builder.CreateString("z"), 8,
                                                             foxglove::NumericType::FLOAT32)});

  // Cast from float to const uint8_t*, required by flatbuffer PointCloud data field
  const uint8_t* data_ptr = reinterpret_cast<const uint8_t*>(cloud_ptr->data.data());

  std::cout << "Cloud, data size: " << data_size << std::endl;

  const auto cloud_flatbuffer =
    foxglove::CreatePointCloud(builder, getFoxgloveTimestamp(cloud_ptr->timestamp.counts).get(),
                               builder.CreateString("base_link"), dummy_pose, stride, fields,
                               builder.CreateVector(data_ptr, data_size));

  builder.Finish(cloud_flatbuffer);
  return SerializedFlatBuffer(builder);
}

bool SensorDataRecorder::writeFoxgloveMsg(const SerializedFlatBuffer msg_data,
                                          const std::string& stream_name,
                                          int64_t sensor_ts)
{
  mcap::Message msg;
  std::cout << " stream name:" << stream_name;
  if (to_channel_id_.find(stream_name) == to_channel_id_.end()) {
    log::warn("Stream name '{}' does not correspond to an existing channel", stream_name);
    return false;
  }

  msg.channelId = to_channel_id_[stream_name];
  msg.logTime = sensor_ts;
  msg.publishTime = sensor_ts; // TODO: get the current timestamp?
  msg.data = msg_data.data;
  msg.dataSize = msg_data.size;

  auto status = writer_.write(msg);
  std::cout << ", msg size:" << msg.dataSize << std::endl; // TODO: remove
  if (!status.ok()) {
    log::warn("Failed to write message from stream {} to mcap. Writer error: {}", stream_name,
              status.message);
    return false;
  }
  return true;
}

void SensorDataRecorder::callback(SensorDataPtr frame)
{
  // convert PerceptionDataFrame to ROS2 schema for writing to file
  std::string stream_name = frame->stream_name.toString();
  if (to_channel_id_.find(stream_name) == to_channel_id_.end()) {
    log::error(
      "Could not associate message channel with its stream name. Stream name is invalid: {}",
      stream_name);
    return;
  }

  SerializedFlatBuffer msg_data;
  if (const auto cloud_ptr = std::dynamic_pointer_cast<PointCloudFrame>(frame); cloud_ptr != nullptr) {
    msg_data = getFoxgloveCloud(cloud_ptr);
    std::cout << "cloud";
  }

  writeFoxgloveMsg(msg_data, stream_name, frame->timestamp.counts);
}
```

I've attached two files in a .zip file:
https://drive.google.com/file/d/1UnQSL7LdYcDjUL2wrcCo_kQfdHgo5qRc/view?usp=sharing
1. points-images: Mcap file with PointClouds, FrameTransforms, and RawImages that runs without the chunk error.
2. points-only-chunk-err: Mcap file with PointClouds, FrameTransforms that triggers the chunk error.

I'm not sure if this an issue with my code or a possible bug. Thank you for your help!

**Comment by defunctzombie** [comment]
Internal tracking ticket: https://linear.app/foxglove/issue/FG-4408/incorrect-chunk-error-with-flatbuffers

**Reply by andrewwongagility** [reply]
Thank you @defunctzombie! Is there any way to track the ticket externally? I'm assuming the link you posted is for Foxglove internal use, but I wasn't sure

**Reply by defunctzombie** [reply]
We will update the github issue once we resolve the ticket. The latest comment on this discussion seems to indicate your file itself is corrupt.

**Comment by achim-k** [comment]
How were you recording the MCAP files? One of two MCAP files that you provided seems to be corrupt, running `mcap doctor` on them  yields the following:

- **points-images.mcap**: No errrors
- **points-only-chunk-err.mcap**: `Doctor command failed: encountered 224 errors`

<details><pre>Header.profile field "flatbuffer" is not a well-known profile.
invalid CRC: c154eb4d != d7bcdebf
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
invalid CRC: 5d255df != 6c6678f5
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Got a Message record for channel: 2 before a channel info.
Got a Message record for channel: 1 before a channel info.
Statistics has message start time 11531378789231, but the minimum message start time is 11531508845491
Statistics has message count 230, but actual number of messages is 220
Doctor command failed: encountered 224 errors
</pre></details>

**Reply by andrewwongagility** [reply]
@achim-k Thank you for your quick reply! 

* I've edited my original post to include the code that writes the pointcloud to the mcap file. Please let me know if you have any questions
* I recorded both files with the same software version. The only difference is that `points-only-chunk-err.mcap` does not include RawImage messages. 

After some more experimentation, **I found that recording both PointCloud and RawImage messages together can still lead to the same chunk error.** Therefore, I believe writing `PointCloud` messages (by itself or with `RawImage` messages) to a mcap file can cause the chunk error. My original post was incorrect.

I'm still new to mcap, but one difference between `PointCloud` and `RawImage` messages are: `RawImage` messages are always the same size, whereas `PointCloud` messages vary based on the number of valid points. I've attached a snippet of the console out below that shows the size in bytes of each message. 

Is it possible that differences between sizes of each PointCloud messages could cause the chunk error? I did try similar code that uses Protobuf to seriailize the messages, and I did not get any chunk errors when using Protobuf


```
transform stream name:tf, msg size:176
Cloud, data size: 38340
cloud stream name:downward-chest-camera/depth/points, msg size:38616

transform stream name:tf, msg size:176
Img size:3072000, msg encoding:rgb8, SensorData format:rgb8 stream name:downward-chest-camera/color/image-rect, msg size:3072104

transform stream name:tf, msg size:176
Cloud, data size: 38640
cloud stream name:downward-chest-camera/depth/points, msg size:38920

transform stream name:tf, msg size:176
Img size:3072000, msg encoding:rgb8, SensorData format:rgb8 stream name:downward-chest-camera/color/image-rect, msg size:3072104

transform stream name:tf, msg size:176
Cloud, data size: 40068
cloud stream name:downward-chest-camera/depth/points, msg size:40344

```

**Reply by achim-k** [reply]
> Is it possible that differences between sizes of each PointCloud messages could cause the chunk error?

This shouldn't be a problem, the mcap writer can handle that. To me it seems more like a concurrency issue. Could it be that multiple threads simultaneously call ` writer_.write(msg)`? That would quite likely cause your mcap file to be corrupt.

**Reply by andrewwongagility** [reply]
@achim-k That is a good point! I believe you're right. From some preliminary testing, locking the callback() function appeared to prevent the error. I'll rework the code and post my solution here. Thanks so much for your help!

**Comment by andrewwongagility** [comment, marked as answer]
After some more debugging, I believe the error is actually caused by how I was using `flatbuffers::FlatBufferBuilder`. At the start of the function `getFoxgloveCloud()`, I construct an instance of `FlatBufferBuilder`. The function returns an object that contains a pointer to the memory owned by `FlatBufferBuilder`.

When the function exits, the memory that was owned by `FlatBufferBuilder` is freed. Therefore, the memory that the object was pointing to is partially or completely overwritten before the memory contents are written to the mcap file. This results in the file corruption mentioned above

One possible solution is returning `FlatBufferBuilder` itself (instead of another object) so the memory is not freed upon returning from the function.

**Summary: chunk error was caused by `FlatBufferBuilder` going out of scope and its memory being freed before the message contents were written to the mcap file, causing the file corruption**

Thanks again to @achim-k and @defunctzombie for all your help!
